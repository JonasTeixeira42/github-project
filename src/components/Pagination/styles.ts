import styled, { css } from 'styled-components'
import { lighten } from 'polished'

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`

export const Button = styled.button`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.white};
    border: 0;
    outline: 0;
    padding: ${theme.spacings.xxsmall} 1.4rem;
    cursor: pointer;
    border-radius: ${theme.border.radius};
    background-color: ${theme.colors.blue};

    &:disabled {
      cursor: not-allowed;
      background-color: ${lighten(0.2, theme.colors.blue)};
    }
  `}
`

export const Page = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xsmall};
  `}
`
