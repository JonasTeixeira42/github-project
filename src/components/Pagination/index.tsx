import React, { useState } from 'react'

import * as S from './styles'

export type PaginationProps = {
  dataLength: number
  itemsPerPage: number
  onChangePage?: (startIndex: number, endIndex: number) => void
}

const Pagination = ({
  dataLength,
  itemsPerPage,
  onChangePage
}: PaginationProps) => {
  const pages = Math.ceil(dataLength / itemsPerPage)
  const [currentPage, setCurrentPage] = useState(1)

  const handleNextPage = () => {
    const startIndex = (currentPage + 1) * itemsPerPage - itemsPerPage
    const endIndex = startIndex + itemsPerPage

    setCurrentPage((page) => page + 1)
    !!onChangePage && onChangePage(startIndex, endIndex)
  }

  const handlePreviousPage = () => {
    const startIndex = (currentPage - 1) * itemsPerPage - itemsPerPage
    const endIndex = startIndex + itemsPerPage

    setCurrentPage((page) => page - 1)
    !!onChangePage && onChangePage(startIndex, endIndex)
  }

  return (
    <S.Wrapper>
      <div>
        <S.Button disabled={currentPage <= 1} onClick={handlePreviousPage}>
          Previous
        </S.Button>
      </div>
      <S.Page>{`${currentPage} of ${pages}`}</S.Page>
      <div>
        <S.Button disabled={currentPage === pages} onClick={handleNextPage}>
          Next
        </S.Button>
      </div>
    </S.Wrapper>
  )
}

export default Pagination
