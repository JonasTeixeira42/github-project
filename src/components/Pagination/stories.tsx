import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'

import Pagination, { PaginationProps } from '.'

export default {
  title: 'Pagination',
  component: Pagination,
  args: {
    dataLength: 20,
    itemsPerPage: 10
  },
  argTypes: {
    nextPage: { action: 'nextPage' },
    previousPage: { action: 'previousPage' }
  }
} as Meta

export const Default: Story<PaginationProps> = (args) => (
  <Pagination {...args} />
)
