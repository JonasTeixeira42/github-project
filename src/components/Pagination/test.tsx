import React from 'react'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderWithTheme } from 'utils/tests/helpers'

import Pagination from '.'

describe('<Pagination />', () => {
  it('should render the Pagination correctly', () => {
    const { container } = renderWithTheme(
      <Pagination dataLength={20} itemsPerPage={10} />
    )

    expect(
      screen.getByRole('button', { name: /previous/i })
    ).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /next/i })).toBeInTheDocument()
    expect(screen.getByText('1 of 2')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should change to the next page when clicking', () => {
    const onClick = jest.fn()
    renderWithTheme(
      <Pagination dataLength={20} itemsPerPage={10} onChangePage={onClick} />
    )

    const button = screen.getByRole('button', { name: /next/i })
    userEvent.click(button)

    expect(onClick).toHaveBeenCalledTimes(1)
    expect(onClick).toHaveBeenCalledWith(10, 20)
  })

  it('should change to the previous page when clicking', () => {
    const onClick = jest.fn()
    renderWithTheme(
      <Pagination dataLength={20} itemsPerPage={10} onChangePage={onClick} />
    )

    const nextButton = screen.getByRole('button', { name: /next/i })
    const previousButton = screen.getByRole('button', { name: /previous/i })
    userEvent.click(nextButton)
    userEvent.click(previousButton)

    expect(onClick).toHaveBeenCalledTimes(2)
    expect(onClick).toHaveBeenCalledWith(0, 10)
  })

  it('should be disabled when the users cant change pages', () => {
    renderWithTheme(<Pagination dataLength={10} itemsPerPage={10} />)

    const nextButton = screen.getByRole('button', { name: /next/i })
    const previousButton = screen.getByRole('button', { name: /previous/i })

    expect(nextButton).toBeDisabled()
    expect(previousButton).toBeDisabled()
  })
})
