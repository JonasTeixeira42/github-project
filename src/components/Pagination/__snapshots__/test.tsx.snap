// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render the Pagination correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  font-size: 1.6rem;
  color: #ffffff;
  border: 0;
  outline: 0;
  padding: 0.8rem 1.4rem;
  cursor: pointer;
  border-radius: 0.4rem;
  background-color: #3799ff;
}

.c1:disabled {
  cursor: not-allowed;
  background-color: #9dcdff;
}

.c2 {
  font-size: 1.4rem;
}

<div
  class="c0"
>
  <div>
    <button
      class="c1"
      disabled=""
    >
      Previous
    </button>
  </div>
  <p
    class="c2"
  >
    1 of 2
  </p>
  <div>
    <button
      class="c1"
    >
      Next
    </button>
  </div>
</div>
`;
